ragna/deploy/_ui/components/metadata_filters_builder.py

```diff
@@ -0,0 +1,248 @@
+
+        self.key_select = pn.widgets.Select.from_param(
+            self.param.key,
+            name="",
+            css_classes=["metadata-filter-row", "metadata-filter-key"],
+            disabled=key_select_disabled,
+        )
+
+        self.operator_select = pn.widgets.Select.from_param(
+            self.param.operator,
+            name="",
+            css_classes=["metadata-filter-row", "metadata-filter-operator"],
+            disabled=True,
+        )
+
+        self.value_select = pn.widgets.Select.from_param(
+            self.param.value,
+            name="",
+            css_classes=["metadata-filter-row", "metadata-filter-value"],
+            disabled=True,
+        )
+
+        self.delete_button = pn.widgets.ButtonIcon(
+            icon="trash",
+            width=25,
+            height=25,
+            css_classes=["metadata-filter-row", "metadata-filter-delete"],
+        )
+        self.delete_button.on_click(on_delete_callback)
+
+    @staticmethod
+    def set_css(widget, valid=True):
+        if valid:
+            widget.css_classes = [
+                c for c in widget.css_classes if c != "metadata-filter-row-error"
+            ]
+        else:
+            widget.css_classes = widget.css_classes + ["metadata-filter-row-error"]
+
+    @param.depends("key", watch=True)
+    def key_did_change(self):
+        if self.key == "":
+            self.value_select.disabled = True
+            self.operator_select.disabled = True
+            return
+
+        self.value_select.disabled = False
+        self.operator_select.disabled = False
+        self.param.value.objects = [""] + self.valid_key_value_pairs[self.key]
+
+    def is_empty(self):
+        return self.key == "" and self.operator == "" and self.value == ""
+
+    def validate(self):
+        if self.key == "":
+            self.set_css(self.key_select, valid=False)
+        else:
+            self.set_css(self.key_select)
+
+        if self.operator == "":
+            self.set_css(self.operator_select, valid=False)
+        else:
+            self.set_css(self.operator_select)
+
+        if self.value == "":
+            self.set_css(self.value_select, valid=False)
+        else:
+            self.set_css(self.value_select)
+
+        # if any attribute is "" then return False
+        return not (self.key == "" or self.operator == "" or self.value == "")
+
+    def construct_metadata_filter(self):
+        # probably not required, but just in case
+        if self.key == "" or self.operator == "" or self.value == "":
+            return None
+
+        return MetadataFilter(MetadataOperator[self.operator], self.key, self.value)
+
+    def __panel__(self):
+        return pn.Row(
+            self.key_select,
+            self.operator_select,
+            self.value_select,
+            self.delete_button,
+        )
+
+
+class MetadataFiltersBuilder(pn.viewable.Viewer):
+    filter_rows = param.List([])
+    corpus_names = param.List([])
```

Comment:
Required parameters as every corpus name will have a potentially different set of key-values and so we need to refresh the values.